### ISSUE TITLE
Proper method of injecting JavaScript and CSS into content

### ISSUE THREAD

**Issue by black7375** [MrOtherGuy/fx-autoconfig#42]
To explain my use case, I want to create a table of contents on the right in the `about:preferences`.

- https://tscanlin.github.io/tocbot/

![image](https://github.com/MrOtherGuy/fx-autoconfig/assets/25581533/415b044d-2aa0-4774-a3e4-b7c3d89f4100)

To do this, I have to inject JavaScript and CSS into the `about:preferences` page.
I think it's possible with AutoConfig, but I don't know the right way.
(I hope there is no effect on `browser.xhtml`)

**Comment by MrOtherGuy** [comment]
Accessing content documents in general is very tricky and something I try to stay away from.

But if the content document you are trying to access is in the parent process (like about:preferences and about:addons) then you can access them trivially - just include their address in the script header:

```js
// ==UserScript==
// @include       about:preferences
// ==/UserScript==
console.log(document.location.href)
```

This script should only be executed in `about:preferences` - not the main window which is the default.

**Comment by black7375** [comment]
Oh thank you. I'll try and share the code if I can!!

**Comment by florinsdistortedvision** [comment]
@MrOtherGuy sorry to bother, but it seems on the latest Nightly (127) the scripts targeting such pages as about:preferences don't seem to work for me, even that console.log example you have provided. Can you replicate this issue on your end?

**Comment by MrOtherGuy** [comment]
Hmm. Seems I can indeed replicate that. Do you know the exact date when this started to happen?

**Comment by florinsdistortedvision** [comment]
Unfortunately I don't, i have tried your example list time on 115ESR which worked fine, and haven't tried since. I was working on something today for the latest Nightly and it just nothing seemed to work.

**Comment by MrOtherGuy** [comment]
All right, I also tried in Firefox 125 and there it seems to work (although fx-autoconfig version I have on my release install is older, but I kinda doubt that's the issue).

What seems to be happening is that loading about:preferences is not firing observer with topic "domwindowopened". That is the one and only method the loader uses to determine when a new "injection target" is being loaded.

**Comment by florinsdistortedvision** [comment]
Interesting, not sure what Mozilla might've broke since 125. The reason I was using Nightly was to prep for 128ESR release which is coming soon.

**Comment by florinsdistortedvision** [comment]
@MrOtherGuy found the issue, if I use the utils from 0.8.6 version everything is created fine and (console log). However it is broken in the latest utils version 0.9.0

**Comment by MrOtherGuy** [comment]
Yes, I just figured this same as well. So my initial assessment was wrong, scripts are not executed in about:preferences or equivalent not because `domwindowopened` was not observed (it never was), but because in newer versions the DOMWindowOpened *event* listener is removed after the first callback - which is called for the window itself. In earlier versions, the event listener stayed around and was then called for any (main process) document that was later opened inside that window.

I'm pretty sure I stumbled upon some problem that was caused by the old behavior because some script was running unexpectedly, but I cannot remember what that was.

It would be easy and I suppose pretty reasonable to add a new pref that essentially restores the old behavior - although that was never really intentional, it just accidentally worked.

**Comment by angelbruni** [comment]
Hello @MrOtherGuy, I stumbled upon this issue which @florinsdistortedvision mentions my problem.
I was trying to update from `0.8.6` to `0.10.1`, all of my code works, except for the ones in pages such as `about:preferences` and `about:addons`. Was anything ever done about this? Thank you.

**Comment by MrOtherGuy** [comment]
Yes, if you are facing that same issue then you can create new boolean pref `userChromeJS.persistent_domcontent_callback` and set it to `true` and your custom scripts should start to load again.

**Comment by angelbruni** [comment]
Thank you so much!!! I will try this when I wake up later and report back :)

**Comment by angelbruni** [comment]
This works! Thank you so much for this :D
